### ISSUE TITLE
Shipping Calculators Configurable In Admin Shipping Methods

### ISSUE THREAD

**Issue by silverbackdan** [Sylius/Sylius#8176]
There are a number of useful shipping calculators available:
http://docs.sylius.org/en/latest/components/Shipping/calculators.html

It appears in the admin area, shipping methods can only have flat rates, or flat rates per item defined. Apologies if I'm missing something, but is it not possible to have the FlexibleRateCalculator, for example, option available when selecting rate calculations on a shipping method?

I believe this is just a feature request though, but a dev who has been working on this may know more as to if this had been implemented on earlier versions.

**Comment by silverbackdan** [comment]
So I've seen that the calculator is newer than the latest beta release but has not been implemented with a form as far as I'm aware yet. If it's helpful, this is how I've implemented the flexible rate calculator manually into the latest beta release:

1. Of course I just copied the FlexibleRateCalculator class into my own AppBundle to be used instead.

2. The service defined in YML (pointing to the copied class in my AppBundle)
```yaml
sylius.shipping_calculator.flexible_rate:
        class: AppBundle\Component\Shipping\Calculator\FlexibleRateCalculator
        tags:
            - { name: sylius.shipping_calculator, calculator: flexible_rate, form-type: AppBundle\Form\Type\Calculator\ChannelBasedFlexibleRateConfigurationType, label: "sylius.form.shipping_calculator.flexible_rate_configuration.label" }
```

3. This points to a ChannelBasedFlexibleRateConfigurationType. In Sylius this would be moved to the namespace `Sylius\Bundle\CoreBundle\Form\Type\Shipping\Calculator`
```php
<?php
// Namespace would need changing if implemented in Sylius core
namespace AppBundle\Form\Type\Calculator;

use Sylius\Bundle\CoreBundle\Form\Type\ChannelCollectionType;
use Sylius\Component\Core\Model\ChannelInterface;
use Symfony\Component\Form\AbstractType;
use Symfony\Component\OptionsResolver\OptionsResolver;

/**
 * @author Daniel West <daniel@silverback.is>
 */
final class ChannelBasedFlexibleRateConfigurationType extends AbstractType
{
    /**
     * {@inheritdoc}
     */
    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults([
            'entry_type' => FlexibleRateConfigurationType::class,
            'entry_options' => function (ChannelInterface $channel) {
                return [
                    'label' => $channel->getName(),
                    'currency' => $channel->getBaseCurrency()->getCode(),
                ];
            },
        ]);
    }

    /**
     * {@inheritdoc}
     */
    public function getParent()
    {
        return ChannelCollectionType::class;
    }

    /**
     * {@inheritdoc}
     */
    public function getBlockPrefix()
    {
        return 'sylius_channel_based_shipping_calculator_flexible_rate';
    }
}

```
I also defined this as a service (perhaps unnecessary though)
```yaml
sylius.form.type.channel_based_shipping_calculator.per_unit_rate:
        class: AppBundle\Form\Type\Calculator\ChannelBasedFlexibleRateConfigurationType
        tags:
          - { name: form.type }
```

4. The class above then calls FlexibleRateConfigurationType which in Sylius would be in the namespace `Sylius\Bundle\ShippingBundle\Form\Type\Calculator`
```php
<?php
// Namespace would need changing if implemented in Sylius core
namespace AppBundle\Form\Type\Calculator;

use Sylius\Bundle\MoneyBundle\Form\Type\MoneyType;
use Symfony\Component\Form\AbstractType;
use Symfony\Component\Form\Extension\Core\Type\IntegerType;
use Symfony\Component\Form\FormBuilderInterface;
use Symfony\Component\OptionsResolver\OptionsResolver;
use Symfony\Component\Validator\Constraints\NotBlank;
use Symfony\Component\Validator\Constraints\Type;

/**
 * @author Daniel West <daniel@silverback.is>
 */
final class FlexibleRateConfigurationType extends AbstractType
{
    /**
     * {@inheritdoc}
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('first_item_cost', MoneyType::class, [
                'label' => 'sylius.form.shipping_calculator.flexible_rate_configuration.first_item_cost',
                'constraints' => [
                    new NotBlank(['groups' => ['sylius']]),
                    new Type(['type' => 'integer', 'groups' => ['sylius']]),
                ],
                'currency' => $options['currency'],
            ])
            ->add('additional_item_cost', MoneyType::class, [
                'label' => 'sylius.form.shipping_calculator.flexible_rate_configuration.additional_item_cost',
                'constraints' => [
                    new NotBlank(['groups' => ['sylius']]),
                    new Type(['type' => 'integer', 'groups' => ['sylius']]),
                ],
                'currency' => $options['currency'],
            ])
            ->add('additional_item_limit', IntegerType::class, [
                'label' => 'sylius.form.shipping_calculator.flexible_rate_configuration.additional_item_limit',
                'constraints' => [
                    new NotBlank(['groups' => ['sylius']]),
                    new Type(['type' => 'integer', 'groups' => ['sylius']]),
                ]
            ])
        ;
    }

    /**
     * {@inheritdoc}
     */
    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver
            ->setDefaults([
                'data_class' => null,
            ])
            ->setRequired('currency')
            ->setAllowedTypes('currency', 'string')
        ;
    }

    /**
     * {@inheritdoc}
     */
    public function getBlockPrefix()
    {
        return 'sylius_shipping_calculator_flexible_unit_rate';
    }
}
```

Then finally translations messages.en.yml:
```yaml
sylius:
    form:
        shipping_calculator:
            flexible_rate_configuration:
                label: Flexible rate per item
                first_item_cost: First item shipping cost
                additional_item_cost: Additional item shipping cost
                additional_item_limit: Additional item limit
```

I've got to the stage now where the option is available in the admin. I'll post updates if there are any other adjustments required or if someone does spot an issue with this approach I'd love to know.

Thanks.

**Comment by silverbackdan** [comment]
Updated - calculator needed updating to be in-line with used calculators now.
```php
<?php
// Namespace changed - class was not released yet
namespace AppBundle\Component\Shipping\Calculator;

use Sylius\Component\Core\Exception\MissingChannelConfigurationException;
use Sylius\Component\Core\Model\ShipmentInterface;
use Sylius\Component\Shipping\Calculator\CalculatorInterface;
use Sylius\Component\Shipping\Model\ShipmentInterface as BaseShipmentInterface;
use Webmozart\Assert\Assert;

class FlexibleRateCalculator implements CalculatorInterface
{

    /**
     * {@inheritdoc}
     *
     * @throws MissingChannelConfigurationException
     */
    public function calculate(BaseShipmentInterface $subject, array $configuration)
    {
        Assert::isInstanceOf($subject, ShipmentInterface::class);

        $channelCode = $subject->getOrder()->getChannel()->getCode();

        if (!isset($configuration[$channelCode])) {
            throw new MissingChannelConfigurationException(sprintf(
                'Channel %s has no amount defined for shipping method %s',
                $subject->getOrder()->getChannel()->getName(),
                $subject->getMethod()->getName()
            ));
        }

        $firstItemCost = $configuration[$channelCode]['first_item_cost'];
        $additionalItemCost = $configuration[$channelCode]['additional_item_cost'];
        $additionalItemLimit = $configuration[$channelCode]['additional_item_limit'];

        $totalItems = $subject->getShippingUnitCount();
        $additionalItems = $totalItems - 1;
        if (0 !== $additionalItemLimit) {
            $additionalItems = $additionalItemLimit >= $additionalItems ? $additionalItems : $additionalItemLimit;
        }
        $finalCost = (int)($firstItemCost + ($additionalItems * $additionalItemCost));

        return $finalCost;
    }

    /**
     * {@inheritdoc}
     */
    public function getType()
    {
        return 'flexible_rate';
    }
}
```

Only issue is now the method `$subject->getShippingUnitCount();` still returns 2 instead of 1 if an order has an item in the basket with a quantity of 2, and this is reduced to 1. You have to update the basket twice for the shipping total to update then. Should this be raised as a separate issue?

**Comment by silverbackdan** [comment]
Final comment, I've hacked a workaround but I'm sure the shipping units should be updated after the cart has been updated. For now, I've updated the shipping calculations to do this instead of using getShippingUnitCount()
```php
        $totalItems = 0;//$subject->getShippingUnitCount();
        foreach( $subject->getOrder()->getItems() as $item )
        {
            if( $item->getVariant()->isShippingRequired() )
            {
                $totalItems+=$item->getQuantity();
            }
        }
```

**Comment by silverbackdan** [comment]
Sorry, there is an additional issue, postage doesn't appear to be calculated when I add an item to the basket now, only when I update. So it'll show as free postage until I update the basket... Am I missing something here?

**Comment by silverbackdan** [comment]
@pjedrzejewski Are you able to point me in the right direction as to why in the beta 2 version if I add products to my basket the shipping appears as £0 until a further save? Do you happen to remember any pull requests or issues that may have looked at this because I cannot replicate on your demo site.

**Comment by lchrusciel** [comment]
Are you willing to open a PR with fixes to our documentation? We should remove obsolete calculators and update DIY guide. Also, a package of an additional shipping calculators is a good idea for a plugin, so if you wish to maintain your very  own package be my guest :)

**Comment by silverbackdan** [comment]
Sure, I can't do it immediately but at least this is here for reference for the time being. I'll get onto it as soon as I can though.

**Comment by marcotisi** [comment]
Hi Sylius developers!
I saw in the documentation that there were additional Shipping Calculators, but they has been removed in the commit https://github.com/Sylius/Shipping/commit/92fad5bf1dcf0bb4cce181e517e56a0f0ba0e645 .
Can you please tell me why? Do you think they will never be used or there are another reasons?
Thank you!

**Comment by ecornelisse** [comment]
I too wish to hear more about the additional shipping calculators based on size and weight

**Comment by Zales0123** [comment]
@marcotisi @ecornelisse Hello! These calculators was removed some time ago, as they were not properly tested and we decided that there is no point for now to maintain them. We focused on making it as easy as possible to implement them on your own.
We have documentation about implementing custom shipping calculators: http://docs.sylius.org/en/latest/components_and_bundles/bundles/SyliusShippingBundle/custom_calculators.html
Sadly, it's quite updated (see https://github.com/Sylius/Sylius/issues/7929 and https://github.com/Sylius/Sylius/issues/8139) but as I know there are some efforts taken to fix this problems.
Nevertheless, I hope that this documentation + related issues will help you both in implementing your own, nice and shiny calculators ;)
